### PR TITLE
[SPARK-32986][SQL] Add bucketed scan info in query plan of data source v1

### PR DESCRIPTION
**Pull request by c21**
### What changes were proposed in this pull request?

As a followup from discussion in https://github.com/apache/spark/pull/29804#discussion_r493229395 , currently the query plan for data source v1 scan operator - `FileSourceScanExec` has no information to indicate whether the table is read as bucketed table or not. And if table not read as bucketed table, what's the reason behind it. Add this info into `FileSourceScanExec` physical query plan output, can help users and developers understand query plan more easily without spending a lot of time debugging why table is not read as bucketed table.

### Why are the changes needed?

Help users and developers debug query plan for bucketed table.

### Does this PR introduce _any_ user-facing change?

The added `Bucketed` information in physical query plan when reading bucketed table.
Note for reading non-bucketed table, the query plan stays same and nothing is changed.

Example:

```
Seq((1, 2), (2, 3)).toDF("i", "j").write.bucketBy(8, "i").saveAsTable("t1")
Seq(2, 3).toDF("i").write.bucketBy(8, "i").saveAsTable("t2")
val df1 = spark.table("t1")
val df2 = spark.table("t2")
df1.join(df2, df1("i") === df2("i"))
```

```
AdaptiveSparkPlan isFinalPlan=false
+- SortMergeJoin [i#20], [i#24], Inner
   :- Sort [i#20 ASC NULLS FIRST], false, 0
   :  +- Filter isnotnull(i#20)
   :     +- FileScan parquet default.t1[i#20,j#21] Batched: true, Bucketed: true, DataFilters: [isnotnull(i#20)], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/Users/chengsu/spark/sql/core/spark-warehouse/org.apache.spark.sq..., PartitionFilters: [], PushedFilters: [IsNotNull(i)], ReadSchema: struct<i:int,j:int>, SelectedBucketsCount: 8 out of 8
   +- Sort [i#24 ASC NULLS FIRST], false, 0
      +- Filter isnotnull(i#24)
         +- FileScan parquet default.t2[i#24] Batched: true, Bucketed: true, DataFilters: [isnotnull(i#24)], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/Users/chengsu/spark/sql/core/spark-warehouse/org.apache.spark.sq..., PartitionFilters: [], PushedFilters: [IsNotNull(i)], ReadSchema: struct<i:int>, SelectedBucketsCount: 8 out of 8
```

### How was this patch tested?

Added unit test in `ExplainSuite.scala`.

### CODE REVIEW

**Comment by c21** [comment]
@cloud-fan and @maropu could you help take a look when you have time? Thanks!

**Review by dongjoon-hyun — APPROVED** [review]
It looks good to me.

File: sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala [hunk 1]
```diff
@@ -677,6 +677,33 @@ class ExplainSuiteAE extends ExplainSuiteHelper with EnableAdaptiveExecutionSuit
...
+    withTable("t1", "t2") {
+      Seq((1, 2), (2, 3)).toDF("i", "j").write.bucketBy(8, "i").saveAsTable("t1")
+      Seq(2, 3).toDF("i").write.bucketBy(8, "i").saveAsTable("t2")
+      val df1 = spark.table("t1")
+      val df2 = spark.table("t2")
+
+      withSQLConf(SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "0") {
+        checkKeywordsExistsInExplain(
+          df1.join(df2, df1("i") === df2("i")),
+          "Bucketed: true" :: Nil: _*)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala:691]: nit: `"Bucketed: true" :: Nil: _*` -> `"Bucketed: true"`?

File: sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala [hunk 2]
```diff
@@ -677,6 +677,33 @@ class ExplainSuiteAE extends ExplainSuiteHelper with EnableAdaptiveExecutionSuit
...
+      }
+
+      withSQLConf(SQLConf.BUCKETING_ENABLED.key -> "false") {
+        checkKeywordsExistsInExplain(
+          df1.join(df2, df1("i") === df2("i")),
+          "Bucketed: false (disabled by configuration)" :: Nil: _*)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala:697]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala [hunk 3]
```diff
@@ -677,6 +677,33 @@ class ExplainSuiteAE extends ExplainSuiteHelper with EnableAdaptiveExecutionSuit
...
+      }
+
+      checkKeywordsExistsInExplain(df1, "Bucketed: false (disabled by query planner)" :: Nil: _*)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala:700]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala [hunk 4]
```diff
@@ -677,6 +677,33 @@ class ExplainSuiteAE extends ExplainSuiteHelper with EnableAdaptiveExecutionSuit
...
+
+      checkKeywordsExistsInExplain(
+        df1.select("j"),
+        "Bucketed: false (bucket column(s) not read)" :: Nil: _*)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala:704]: ditto

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala (hunk 1 above)
Comment (c21) [reply sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala]: yeah, updated for this and other places.

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 3.3.0 according to the issue type, `Improvement`.

**Comment by dongjoon-hyun** [comment]
Thank you, @c21 and @cloud-fan .

**Comment by c21** [comment]
Thank you @dongjoon-hyun and @cloud-fan for review!

**Pull request closed without merging**